---
title: "Plot.ly Volcano Plot Example"
author: "Stephen Kelly"
date: "9/24/2016"
output: 
  html_document: 
    fig_height: 8
    fig_width: 8
    keep_md: yes
    number_sections: true
    code_folding: show
    toc: false
    toc_float: true
citation_package: natbib
bibliography: references.bib
biblio-style: apsr
---

In this example, I will demonstrate how to use gene differential binding data to create a volcano plot using R and Plot.ly. This dataset was generated by DiffBind during the analysis of a ChIP-Seq experiment. Each entry represents a bound peak that was differentially expressed between groups of samples.


First, install any libraries you might not have, load the Plot.ly library, and load the file containing your differential binding data.


```{r}
# dependencies:
# install.packages("ggplot2")
# install.packages("gridExtra")
# install.packages("plotly")
suppressPackageStartupMessages(library("plotly"))

# path to the DiffBind table with genes
input_file <- "../sample_data/diff_bind.Treatment4-ChIPSeq-vs-Control-ChIPSeq.p100.csv"

# read the file into a dataframe
diff_df <- read.delim(file = input_file,header = TRUE,sep = ',')

# check some attributes of the data
colnames(diff_df)

dim(diff_df)
```


This table has a lot of information, but for this plot we only need the following:

- Gene names
- Fold change value
- Statistical significance (e.g. p value, adjusted p value, etc.)

In this case we are using the False Discovery Rate as our signficance value.

```{r}
# keep only the fields needed for the plot
# FDR = false discovery rate = adjusted p value = significance 
diff_df <- diff_df[c("external_gene_name", "Fold", "FDR")]

# preview the dataset; data required for the plot
head(diff_df)
```


In order to color the points on the plot easier, we will add an extra column to the data to mark whether each entry passes our cutoffs for Fold change and Significance. Our criteria for this experiment were:

- Fold change > 1.5x change
- FDR < 0.05 (similar to a p value of <0.05)


```{r}
# add a grouping column; default value is "not significant"
diff_df["group"] <- "NotSignificant"

# for our plot, we want to highlight 
# FDR < 0.05 (significance level)
# Fold Change > 1.5

# change the grouping for the entries with significance but not a large enough Fold change
diff_df[which(diff_df['FDR'] < 0.05 & abs(diff_df['Fold']) < 1.5 ),"group"] <- "Significant"

# change the grouping for the entries a large enough Fold change but not a low enough p value
diff_df[which(diff_df['FDR'] > 0.05 & abs(diff_df['Fold']) > 1.5 ),"group"] <- "FoldChange"

# change the grouping for the entries with both significance and large enough fold change
diff_df[which(diff_df['FDR'] < 0.05 & abs(diff_df['Fold']) > 1.5 ),"group"] <- "Significant&FoldChange"
```

We would also like to label the top 10 genes; the ones with the highest positive or negative fold changes and FDR values. 

```{r}

# Find and label the top peaks..
top_peaks <- diff_df[with(diff_df, order(Fold, FDR)),][1:5,]
top_peaks <- rbind(top_peaks, diff_df[with(diff_df, order(-Fold, FDR)),][1:5,])


# Add gene labels to the plot
# Single Gene Annotation example
# m <- diff_df[with(diff_df, order(Fold, FDR)),][1,]
# a <- list(
#   x = m[["Fold"]],
#   y = -log10(m[["FDR"]]),
#   text = m[["external_gene_name"]],
#   xref = "x",
#   yref = "y",
#   showarrow = TRUE,
#   arrowhead = 7,
#   ax = 20,
#   ay = -40
# )

# Add gene labels for all of the top genes we found
# here we are creating an empty list, and filling it with entries for each row in the dataframe
# each list entry is another list with named items that will be used by Plot.ly
a <- list()
for (i in seq_len(nrow(top_peaks))) {
  m <- top_peaks[i, ]
  a[[i]] <- list(
    x = m[["Fold"]],
    y = -log10(m[["FDR"]]),
    text = m[["external_gene_name"]],
    xref = "x",
    yref = "y",
    showarrow = TRUE,
    arrowhead = 0.5,
    ax = 20,
    ay = -40
  )
}

```

Now we can make the plot

```{r}
# make the Plot.ly plot
p <- plot_ly(data = diff_df, x = Fold, y = -log10(FDR), text = external_gene_name, mode = "markers", color = group) %>% 
  layout(title ="Volcano Plot") %>%
  layout(annotations = a)
p
```


```{r}
# to save plot to a HTML file:
# htmlwidgets::saveWidget(as.widget(p), "graph.html")

# System Information
sessionInfo()
```



# References

https://plot.ly/r/offline/  
https://plot.ly/r/  
https://plot.ly/r/text-and-annotations/  
https://plot.ly/r/reference/#Layout_and_layout_style_objects  
http://www.gettinggeneticsdone.com/2014/05/r-volcano-plots-to-visualize-rnaseq-microarray.html  
